woocommerce/woocommerce-gutenberg-products-block
# Add preliminary extensibility points for Product Bundles support (Cart block)

This PR introduces some new extensibility points to add Cart/Checkout Blocks support in Product Bundles, Composite Products, and likely other extensions, such as Min/Max Quantities.


Ref #3895 

This doesn't close #3895 yet - it includes some of the basic extensibility points we've identified after studying the _documented_ stories and scenarios of PB. As the development of these integrations moves forward, it is likely that more missing bits and pieces will appear.

Ideally, some of these changes should/could have been opened in individual PRs, however:

- To make reviewing easier I have clearly separated each change in its own commit.
- Since all of these changes are required by #3895 it makes a lot of sense to review them together. With the exception of https://github.com/woocommerce/woocommerce-gutenberg-products-block/pull/4241/commits/af91ae8ba159c496da3661c215854eda4ab19118 all other commits are probably low-risk and rather small in scope. Let me know if you’d rather see a separate PR for each commit.

### Screenshots

<img width="1125" alt="Image 2021-05-19 at 5 34 57 p m" src="https://user-images.githubusercontent.com/1783726/118831630-97855600-b8c8-11eb-8906-c7f9ada769e7.png">

### How to test the changes in this Pull Request:

1. Clone PB and checkout branch `issue-758` (ping me for repo access). Alternatively, I can send a link to a pre-release version (still in flux).
2. Create a Bundle like [this](https://user-images.githubusercontent.com/1783726/118832114-04005500-b8c9-11eb-97af-2978861bc3fc.png). Leave the **General > Regular/Sale Price** fields empty, but make sure that at least one bundled item is **Priced Individually**.
3. Add the bundle to your cart.

At this point:

- Bundled line items should appear indented.
- Bundled product quantities should not be freely editable and will stay in sync with "container" quantity.
- When the "container" item quantity is changed, bundled item quantities should be updated to reflect the new quantities returned by the API. This has been discussed in #3895.
- The "container" item should list the aggregated bundled product prices + subtotals.

A couple notes:

> When the "container" item quantity is changed, bundled item quantities should be updated to reflect the new quantities returned by the API

See https://github.com/woocommerce/woocommerce-gutenberg-products-block/commit/af91ae8ba159c496da3661c215854eda4ab19118 -- I am not entirely confident with my approach here. My goal here was to make sure that the quantity passed down from that API in `lineItem` is fed into the `QuantitySelector` component unless there's a pending quantity update triggered by user input. See https://github.com/woocommerce/woocommerce-gutenberg-products-block/pull/4241/files#diff-d4fa4fa6c42594eee9e65823c4d7fb7c92ba45afa28d4696baced7131bbc6770R303

To know whether "there's a pending quantity update triggered by user input" some modifications to the `QuantitySelector` component were necessary, because there's some built-in throttling (debouncing) there that prevents `isPending.quantity` from returning `true` immediately after a quantity increment/decrement: https://github.com/woocommerce/woocommerce-gutenberg-products-block/pull/4241/files#diff-164a1373b934919b2a907a1fede87c8c91c3298de7ca55a2bd35ed98a814d3c0R144

This part needs a solid review. 

> Bundled product quantities should not be freely editable.

Min/step handling in the quantity selector component had to go all the way to the API. This is a vital requirement that in my mind shouldn't be left to third parties to implement (this needs to be supported at component level anyway). Note that I haven't added any filters in the API schema class yet (maybe I should?): https://github.com/woocommerce/woocommerce-gutenberg-products-block/pull/4241/files#diff-4a0ace31a5843c5b16d88d6081f20dee749c9d5309dff0b7c0504bd1482217dbR327-R328

cc @nerrad 

### Changelog

> Added 'cartItemClass' filter for cart items.
> Introduced product title formatting using the 'productNameFormat' filter.
> Added cart item min quantity and quantity step support in the store API and quantity selector component.
> Wrapped cart item product td elements in div for greater styling flexibility.
